File: tools/frontfps/fpquat_c/fpquat.c
/* SPDX-License-Identifier: AGPL-3.0-only
 * Copyright (C) 2026 Daniel J. Dillberg
 *
 * fpquat — SECOND PLACEMENT (C99). Independent build of the Q32.32 rotation
 * substrate (URDRFPQ1, frontfps Stage 2): quaternion mul/norm2/normalize/rotate/
 * nlerp and the frozen rsqrt recipe isqrt(2^96 // x) (Newton integer isqrt with
 * final adjustment), on the FROZEN FIELDFP laws — ONE = 2^32, round-to-nearest
 * ties-away division, i64 REFUSE ceiling (products in __int128, never wrapped).
 * Reproduces the reference battery digest bit-for-bit, twice, checks the rsqrt
 * inequality law and the refusal canaries, and with --defect computes the wrap64
 * battery (the port hazard from the rigidity note: 'a naive i64 multiply wraps
 * and diverges') which MUST diverge from the golden.
 * Build:  cc -O2 -std=c99 fpquat.c -o fpquat && ./fpquat        (and --defect)
 */
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdint.h>

typedef __int128 i128;
typedef unsigned __int128 u128;
static const int64_t ONE = (int64_t)1 << 32;
static const i128 IMAX = ((i128)1 << 63) - 1;
static const int64_t COMP_MAX = (int64_t)1 << 61;
static const char *GOLDEN = "3f4aa0d172713a4bf26433c19e211fbd52e474bbae603ce0c665d145412b7e7a";

/* ---- SHA-256 (own implementation, house pattern) ---------------------------------- */
static uint32_t rotr(uint32_t x, int n){ return (x>>n)|(x<<(32-n)); }
static const uint32_t K[64]={
    0x428a2f98,0x71374491,0xb5c0fbcf,0xe9b5dba5,0x3956c25b,0x59f111f1,0x923f82a4,0xab1c5ed5,
    0xd807aa98,0x12835b01,0x243185be,0x550c7dc3,0x72be5d74,0x80deb1fe,0x9bdc06a7,0xc19bf174,
    0xe49b69c1,0xefbe4786,0x0fc19dc6,0x240ca1cc,0x2de92c6f,0x4a7484aa,0x5cb0a9dc,0x76f988da,
    0x983e5152,0xa831c66d,0xb00327c8,0xbf597fc7,0xc6e00bf3,0xd5a79147,0x06ca6351,0x14292967,
    0x27b70a85,0x2e1b2138,0x4d2c6dfc,0x53380d13,0x650a7354,0x766a0abb,0x81c2c92e,0x92722c85,
    0xa2bfe8a1,0xa81a664b,0xc24b8b70,0xc76c51a3,0xd192e819,0xd6990624,0xf40e3585,0x106aa070,
    0x19a4c116,0x1e376c08,0x2748774c,0x34b0bcb5,0x391c0cb3,0x4ed8aa4a,0x5b9cca4f,0x682e6ff3,
    0x748f82ee,0x78a5636f,0x84c87814,0x8cc70208,0x90befffa,0xa4506ceb,0xbef9a3f7,0xc67178f2};
static void sha256(const unsigned char *data,size_t len,unsigned char out[32]){
    uint32_t h[8]={0x6a09e667,0xbb67ae85,0x3c6ef372,0xa54ff53a,0x510e527f,0x9b05688c,0x1f83d9ab,0x5be0cd19};
    size_t ml=len+1; while(ml%64!=56) ml++; size_t total=ml+8;
    unsigned char *msg=(unsigned char*)calloc(total,1); memcpy(msg,data,len); msg[len]=0x80;
    uint64_t bl=(uint64_t)len*8; for(int i=0;i<8;i++) msg[total-1-i]=(unsigned char)((bl>>(8*i))&0xFF);
    uint32_t w[64];
    for(size_t off=0;off<total;off+=64){
        for(int i=0;i<16;i++) w[i]=((uint32_t)msg[off+4*i]<<24)|((uint32_t)msg[off+4*i+1]<<16)|((uint32_t)msg[off+4*i+2]<<8)|((uint32_t)msg[off+4*i+3]);
        for(int i=16;i<64;i++){ uint32_t s0=rotr(w[i-15],7)^rotr(w[i-15],18)^(w[i-15]>>3); uint32_t s1=rotr(w[i-2],17)^rotr(w[i-2],19)^(w[i-2]>>10); w[i]=w[i-16]+s0+w[i-7]+s1; }
        uint32_t a=h[0],b=h[1],c=h[2],d=h[3],e=h[4],f=h[5],g=h[6],hh=h[7];
        for(int i=0;i<64;i++){ uint32_t S1=rotr(e,6)^rotr(e,11)^rotr(e,25); uint32_t ch=(e&f)^((~e)&g); uint32_t t1=hh+S1+ch+K[i]+w[i];
            uint32_t S0=rotr(a,2)^rotr(a,13)^rotr(a,22); uint32_t maj=(a&b)^(a&c)^(b&c); uint32_t t2=S0+maj;
            hh=g;g=f;f=e;e=d+t1;d=c;c=b;b=a;a=t1+t2; }
        h[0]+=a;h[1]+=b;h[2]+=c;h[3]+=d;h[4]+=e;h[5]+=f;h[6]+=g;h[7]+=hh;
    }
    free(msg);
    for(int i=0;i<8;i++){ out[4*i]=(unsigned char)(h[i]>>24); out[4*i+1]=(unsigned char)(h[i]>>16); out[4*i+2]=(unsigned char)(h[i]>>8); out[4*i+3]=(unsigned char)h[i]; }
}
static void tohex(const unsigned char h[32],char out[65]){ static const char*hx="0123456789abcdef"; for(int i=0;i<32;i++){ out[2*i]=hx[h[i]>>4]; out[2*i+1]=hx[h[i]&0xF]; } out[64]=0; }

/* ---- the frozen laws --------------------------------------------------------------- */
static int refused = 0;                      /* sticky refusal flag (checked per op batch) */
static int64_t rdiv(i128 p, i128 d){         /* round-to-nearest, ties away from zero */
    i128 r = (p >= 0) ? (2*p + d) / (2*d) : -((2*(-p) + d) / (2*d));
    if (r > IMAX || r < -IMAX) { refused = 1; return 0; }
    return (int64_t)r;
}
static int64_t fin(int64_t v){ if (v > COMP_MAX || v < -COMP_MAX) refused = 1; return v; }
static int64_t w64(i128 v){                  /* two's-complement wrap — DEFECT ONLY */
    return (int64_t)(uint64_t)(u128)v;
}
static int bitlen_u128(u128 n){ int b=0; while(n){ b++; n>>=1; } return b; }
static u128 isqrt_newton(u128 n){
    if (n < 2) return n;
    u128 r = (u128)1 << ((bitlen_u128(n)+1)/2);
    for(;;){ u128 nr = (r + n/r) >> 1; if (nr >= r) break; r = nr; }
    while (r*r > n) r--;
    while ((r+1)*(r+1) <= n) r++;
    return r;
}

typedef struct { int64_t w,x,y,z; } Q4;
typedef struct { int64_t x,y,z; } V3;

static int64_t qnorm2(Q4 q){
    fin(q.w); fin(q.x); fin(q.y); fin(q.z);
    i128 s = (i128)q.w*q.w + (i128)q.x*q.x + (i128)q.y*q.y + (i128)q.z*q.z;
    return rdiv(s, ONE);
}
static int64_t qdot(Q4 p, Q4 q){
    i128 s = (i128)p.w*q.w + (i128)p.x*q.x + (i128)p.y*q.y + (i128)p.z*q.z;
    return rdiv(s, ONE);
}
static Q4 qmul(Q4 p, Q4 q){
    Q4 r;
    r.w = rdiv((i128)p.w*q.w - (i128)p.x*q.x - (i128)p.y*q.y - (i128)p.z*q.z, ONE);
    r.x = rdiv((i128)p.w*q.x + (i128)p.x*q.w + (i128)p.y*q.z - (i128)p.z*q.y, ONE);
    r.y = rdiv((i128)p.w*q.y - (i128)p.x*q.z + (i128)p.y*q.w + (i128)p.z*q.x, ONE);
    r.z = rdiv((i128)p.w*q.z + (i128)p.x*q.y - (i128)p.y*q.x + (i128)p.z*q.w, ONE);
    return r;
}
static int64_t rsqrt_q(int64_t x){
    if (x <= 0) { refused = 1; return 0; }
    u128 n = ((u128)1 << 96) / (u128)x;
    u128 r = isqrt_newton(n);
    if ((i128)r > IMAX) { refused = 1; return 0; }
    return (int64_t)r;
}
static Q4 qnormalize(Q4 q){
    int64_t n2 = qnorm2(q);
    if (n2 <= 0) { refused = 1; Q4 z={0,0,0,0}; return z; }
    int64_t r = rsqrt_q(n2);
    Q4 u;
    u.w = rdiv((i128)q.w*r, ONE); u.x = rdiv((i128)q.x*r, ONE);
    u.y = rdiv((i128)q.y*r, ONE); u.z = rdiv((i128)q.z*r, ONE);
    return u;
}
static V3 vrotate(Q4 q, V3 v){
    int64_t tx = 2*rdiv((i128)q.y*v.z - (i128)q.z*v.y, ONE);
    int64_t ty = 2*rdiv((i128)q.z*v.x - (i128)q.x*v.z, ONE);
    int64_t tz = 2*rdiv((i128)q.x*v.y - (i128)q.y*v.x, ONE);
    V3 o;
    o.x = v.x + rdiv((i128)q.w*tx, ONE) + rdiv((i128)q.y*tz - (i128)q.z*ty, ONE);
    o.y = v.y + rdiv((i128)q.w*ty, ONE) + rdiv((i128)q.z*tx - (i128)q.x*tz, ONE);
    o.z = v.z + rdiv((i128)q.w*tz, ONE) + rdiv((i128)q.x*ty - (i128)q.y*tx, ONE);
    return o;
}
static Q4 qnlerp(Q4 p, Q4 q, int64_t t){
    if (t < 0 || t > ONE) { refused = 1; Q4 z={0,0,0,0}; return z; }
    if (qdot(p, q) < 0) { q.w=-q.w; q.x=-q.x; q.y=-q.y; q.z=-q.z; }
    int64_t li = ONE - t;
    Q4 b;
    b.w = rdiv((i128)p.w*li + (i128)q.w*t, ONE);
    b.x = rdiv((i128)p.x*li + (i128)q.x*t, ONE);
    b.y = rdiv((i128)p.y*li + (i128)q.y*t, ONE);
    b.z = rdiv((i128)p.z*li + (i128)q.z*t, ONE);
    return qnormalize(b);
}
/* wrap64 defect variants (norm2 + qmul only, mirroring the reference defect) */
static int64_t qnorm2_w(Q4 q){
    int64_t s = w64((i128)w64((i128)w64((i128)q.w*q.w) + w64((i128)q.x*q.x))
                  + w64((i128)w64((i128)q.y*q.y) + w64((i128)q.z*q.z)));
    return rdiv((i128)s, ONE);
}
static Q4 qmul_w(Q4 p, Q4 q){
    Q4 r;
    r.w = rdiv((i128)w64((i128)w64((i128)w64((i128)p.w*q.w) - w64((i128)p.x*q.x))
                        - w64((i128)w64((i128)p.y*q.y) + w64((i128)p.z*q.z))), ONE);
    r.x = rdiv((i128)w64((i128)w64((i128)w64((i128)p.w*q.x) + w64((i128)p.x*q.w))
                        + w64((i128)w64((i128)p.y*q.z) - w64((i128)p.z*q.y))), ONE);
    r.y = rdiv((i128)w64((i128)w64((i128)w64((i128)p.w*q.y) - w64((i128)p.x*q.z))
                        + w64((i128)w64((i128)p.y*q.w) + w64((i128)p.z*q.x))), ONE);
    r.z = rdiv((i128)w64((i128)w64((i128)w64((i128)p.w*q.z) + w64((i128)p.x*q.y))
                        - w64((i128)w64((i128)p.y*q.x) - w64((i128)p.z*q.w))), ONE);
    return r;
}

/* ---- battery (mirrors fpquat.py constants and row order exactly) ------------------- */
static unsigned char buf[65536]; static size_t blen;
static void put_tag(const char*t){ size_t n=strlen(t); memcpy(buf+blen,t,n); blen+=n; }
static void put_i64(int64_t v){ uint64_t u=(uint64_t)v; for(int i=0;i<8;i++) buf[blen+i]=(unsigned char)(u>>(8*(7-i))); blen+=8; }

static void battery(int defect, char out_hex[65]){
    const int64_t H2 = ONE/2, Q3 = 3*ONE/4, T3 = ONE/3;
    const Q4 QUATS[4] = {{ONE,0,0,0},{ONE,H2,0,0},{ONE,H2,Q3,-T3},{3*ONE,-2*ONE,ONE,H2}};
    const V3 VECS[4]  = {{ONE,0,0},{0,ONE,0},{H2,Q3,-ONE},{5*ONE,-3*ONE,2*ONE+12345}};
    const int64_t RSQ[10] = {1,2,ONE/4,H2,ONE,2*ONE,3*ONE,10*ONE,((int64_t)1<<48)+7919,COMP_MAX};
    const int64_t NT[4] = {0, ONE/4, H2, ONE};
    blen = 0; memcpy(buf,"URDRFPQ1",8); blen=8;
    for(int i=0;i<10;i++){ put_tag("rsqrt"); put_i64(rsqrt_q(RSQ[i])); }
    for(int i=0;i<4;i++){ put_tag("norm2"); put_i64(defect ? w64((i128)qnorm2_w(QUATS[i])) : qnorm2(QUATS[i])); }
    for(int i=0;i<4;i++) for(int j=0;j<4;j++){
        put_tag("qmul");
        Q4 r = defect ? qmul_w(QUATS[i],QUATS[j]) : qmul(QUATS[i],QUATS[j]);
        if (defect){ put_i64(w64((i128)r.w)); put_i64(w64((i128)r.x)); put_i64(w64((i128)r.y)); put_i64(w64((i128)r.z)); }
        else { put_i64(r.w); put_i64(r.x); put_i64(r.y); put_i64(r.z); }
    }
    Q4 units[4];
    for(int i=0;i<4;i++){ units[i]=qnormalize(QUATS[i]); put_tag("normalize");
        put_i64(units[i].w); put_i64(units[i].x); put_i64(units[i].y); put_i64(units[i].z); }
    for(int i=0;i<4;i++) for(int j=0;j<4;j++){
        put_tag("rotate"); V3 r=vrotate(units[i],VECS[j]); put_i64(r.x); put_i64(r.y); put_i64(r.z); }
    for(int i=0;i<4;i++) for(int k=0;k<4;k++){
        put_tag("nlerp"); Q4 r=qnlerp(QUATS[i],QUATS[(i+1)%4],NT[k]);
        put_i64(r.w); put_i64(r.x); put_i64(r.y); put_i64(r.z); }
    unsigned char h[32]; sha256(buf,blen,h); tohex(h,out_hex);
}

int main(int argc, char**argv){
    int defect = (argc>1 && strcmp(argv[1],"--defect")==0);
    char d1[65], d2[65];
    refused = 0;
    battery(defect, d1);
    battery(defect, d2);
    if (refused){ printf("FPQ-REFUSE fired during battery — inadmissible\n"); return 2; }
    if (defect){
        int caught = strcmp(d1, GOLDEN) != 0;
        printf("wrap64 defect digest: %s\n", d1);
        printf(caught ? "URDR-FPQUAT-C: DEFECT CAUGHT (diverges from golden)\n"
                      : "URDR-FPQUAT-C: DEFECT MISSED — vacuous\n");
        return caught ? 0 : 1;
    }
    /* rsqrt inequality law, refusal canaries */
    const int64_t RSQ[10] = {1,2,ONE/4,ONE/2,ONE,2*ONE,3*ONE,10*ONE,((int64_t)1<<48)+7919,COMP_MAX};
    int law = 1;
    for(int i=0;i<10;i++){ u128 r=(u128)rsqrt_q(RSQ[i]);
        if (!( r*r*(u128)RSQ[i] <= ((u128)1<<96) && (r+2)*(r+2)*(u128)RSQ[i] > ((u128)1<<96) )) law=0; }
    int canaries = 0;
    refused=0; rsqrt_q(0);        if (refused) canaries++;
    refused=0; rsqrt_q(-ONE);     if (refused) canaries++;
    refused=0; fin(COMP_MAX+1);   if (refused) canaries++;
    { Q4 z={0,0,0,0}; refused=0; qnormalize(z); if (refused) canaries++; }
    { refused=0; Q4 a={ONE,0,0,0},b={ONE,ONE/2,0,0}; qnlerp(a,b,ONE+1); if (refused) canaries++; }
    int twice = strcmp(d1,d2)==0, golden = strcmp(d1,GOLDEN)==0;
    printf("battery digest: %s\n", d1);
    printf("twice identical: %s | matches golden: %s | rsqrt law: %s | refusals: %d/5\n",
           twice?"yes":"NO", golden?"yes":"NO", law?"holds":"VIOLATED", canaries);
    if (twice && golden && law && canaries==5){
        printf("URDR-FPQUAT-C: ADMITTED (66-row battery ×2 bit-for-bit, rsqrt law, refusals total)\n");
        return 0;
    }
    printf("URDR-FPQUAT-C: FAILED\n");
    return 1;
}

File: tools/frontfps/fpquat.py
#!/usr/bin/env python3
# SPDX-License-Identifier: AGPL-3.0-only
# Copyright (C) 2026 Daniel J. Dillberg
"""fpquat — the Q32.32 rotation substrate (frontfps Stage 2, URDRFPQ1).

Quaternion operations on the FROZEN FIELDFP fixed-point laws (`tools/physics/
field.py`): radix ONE = 2^32, round-to-nearest ties-away division `_rdiv`, i64
refusal ceiling. Nothing here invents arithmetic — every rounding point reuses
`_rdiv`, and every op is a pinned integer recipe a placement reproduces verbatim
(reproducibility-by-frozen-rounding, the rung-5 philosophy applied to rotation).

Ops (all inputs Q32.32 raw i64, admission bound |raw| ≤ COMP_MAX = 2^61 so every
intermediate provably fits a signed 128-bit integer in every placement):

  fqmul(a,b)        fixed multiply — ONE rounding: _rdiv(a·b, ONE)
  qconj(q)          exact negation of the vector part
  qnorm2(q)         w²+x²+y²+z² summed EXACTLY, then ONE rounding
  qdot(p,q)         mixed products summed exactly, then ONE rounding
  qmul(p,q)         Hamilton product — each component: 4 exact products,
                    exact signed sum, ONE rounding (never 4 roundings)
  rsqrt(x)          reciprocal square root — the frozen integer recipe
                    isqrt(2^96 // x): Newton integer isqrt with final
                    adjustment; NO iteration count to tune, NO float ever;
                    result within 2 ulp of 2^48/√x (proved by inequality,
                    not sampled: r²·x ≤ 2^96 < (r+2)²·x)
  qnormalize(q)     componentwise fqmul(c, rsqrt(qnorm2(q)))
  vrotate(q,v)      v + w·t + u×t with t = 2(u×v); each cross component and
                    each w·t component takes ONE rounding, in the written order
  qnlerp(p,q,t)     shortest-path (qdot < 0 ⇒ negate q, exact) normalized lerp:
                    per component ONE rounding of p·(ONE−t) + q·t, then
                    qnormalize; t ∈ [0, ONE] refused outside — nlerp needs no
                    trigonometry, which is why it is Stage 2 and slerp is not

Refusals are total and typed (`FPQ-REFUSE`): non-integer input, |raw| beyond
COMP_MAX, i64 overflow on any result, rsqrt of x ≤ 0, normalize of a zero/
degenerate quaternion, nlerp t outside [0, ONE]. No wrap, no clamp, no guess.

THE DEFECTS SHIP IN-MODULE (gate non-vacuity): `battery_digest_defect_truncdiv`
(floor division instead of the frozen round-to-nearest — the same law the field
stage proves) and `battery_digest_defect_wrap64` (quaternion products summed in
wrapping 64-bit arithmetic — the porting hazard this repo has actually hit:
the rigidity port note reads 'a naive i64 multiply wraps and diverges'; the
battery's large quaternion exists precisely to force that wrap).

GRADE (Stage 2, reference placement): maturity IMPLEMENTED; evidence MEASURED via
the `frontfps_quat` gate stage + `tests/test_fpquat.py` (battery golden ×2, the
rsqrt inequality law, normalize/rotate bounds, refusal canaries, both defects
diverging). Cross-placement: C99 placement in `fpquat_c/` (self-verified where a
gcc host exists); Rust placement in `fpquat_rs/` SPECULATIVE until a named host
prints ADMITTED twice with the defect caught. `does_not_show`: anything about
angles, slerp, or animation — that is Stage 3; and no real-time performance claim.
Falsifier: any independent implementation of this docstring disagreeing with the
pinned battery digest.
"""
import hashlib
import math
import os
import sys

_HERE = os.path.dirname(os.path.abspath(__file__))
_PHYS = os.path.join(os.path.dirname(_HERE), "physics")
if _PHYS not in sys.path:
    sys.path.insert(0, _PHYS)

from field import ONE, _rdiv  # the FROZEN substrate laws (FIELDFP)  # noqa: E402

IMAX = (1 << 63) - 1
COMP_MAX = 1 << 61          # op-input admission bound (i128 headroom, proved in docstring)
MAGIC = b"URDRFPQ1"


class FpqError(Exception):
    def __init__(self, message):
        super().__init__(f"FPQ-REFUSE: {message}")
        self.code = "FPQ-REFUSE"


def _fg(v):
    if v > IMAX or v < -IMAX:
        raise FpqError(f"i64 overflow ({v})")
    return v


def _fin(v, what):
    if not isinstance(v, int) or isinstance(v, bool):
        raise FpqError(f"{what}={v!r} is not an integer raw Q32.32 word")
    if v > COMP_MAX or v < -COMP_MAX:
        raise FpqError(f"{what}={v} beyond COMP_MAX=±2^61 (admission bound)")
    return v


def _quat(q, what):
    if not (isinstance(q, tuple) and len(q) == 4):
        raise FpqError(f"{what} is not a (w,x,y,z) tuple")
    return tuple(_fin(c, f"{what}[{i}]") for i, c in enumerate(q))


def _vec(v, what):
    if not (isinstance(v, tuple) and len(v) == 3):
        raise FpqError(f"{what} is not an (x,y,z) tuple")
    return tuple(_fin(c, f"{what}[{i}]") for i, c in enumerate(v))


# ---- the frozen integer isqrt (Newton + final adjustment) ----------------------------
def _isqrt_newton(n):
    """Exact floor integer sqrt by Newton's method — THE recipe every placement
    ports (u128 arithmetic suffices for every admitted input). Equality with the
    mathematical isqrt is a falsifier, not a hope (tests assert == math.isqrt)."""
    if n < 0:
        raise FpqError(f"isqrt of negative ({n})")
    if n < 2:
        return n
    r = 1 << ((n.bit_length() + 1) // 2)
    while True:
        nr = (r + n // r) // 2
        if nr >= r:
            break
        r = nr
    # belt-and-braces adjustment: provably unreachable for THIS stop rule, kept
    # so a mis-ported Newton variant still lands exactly on floor(√n)
    while r * r > n:
        r -= 1
    while (r + 1) * (r + 1) <= n:
        r += 1
    return r


def _w64(v):
    """Two's-complement i64 wrap — the arithmetic a careless port gets for free.
    Exists ONLY for the defect below; the real ops never wrap, they refuse."""
    return ((v + (1 << 63)) & ((1 << 64) - 1)) - (1 << 63)


# ---- ops (every rounding point is _rdiv; every recipe is the spec) --------------------
def fqmul(a, b, _div=_rdiv):
    return _fg(_div(_fin(a, "a") * _fin(b, "b"), ONE))


def qconj(q):
    w, x, y, z = _quat(q, "q")
    return (w, -x, -y, -z)


def qnorm2(q, _div=_rdiv):
    w, x, y, z = _quat(q, "q")
    return _fg(_div(w * w + x * x + y * y + z * z, ONE))


def qdot(p, q, _div=_rdiv):
    pw, px, py, pz = _quat(p, "p")
    qw, qx, qy, qz = _quat(q, "q")
    return _fg(_div(pw * qw + px * qx + py * qy + pz * qz, ONE))


def qmul(p, q, _div=_rdiv):
    pw, px, py, pz = _quat(p, "p")
    qw, qx, qy, qz = _quat(q, "q")
    return (_fg(_div(pw * qw - px * qx - py * qy - pz * qz, ONE)),
            _fg(_div(pw * qx + px * qw + py * qz - pz * qy, ONE)),
            _fg(_div(pw * qy - px * qz + py * qw + pz * qx, ONE)),
            _fg(_div(pw * qz + px * qy - py * qx + pz * qw, ONE)))


def rsqrt(x):
    _fin(x, "x")
    if x <= 0:
        raise FpqError(f"rsqrt domain: x={x} ≤ 0")
    return _fg(_isqrt_newton((1 << 96) // x))


def qnormalize(q, _div=_rdiv):
    n2 = qnorm2(q, _div)
    if n2 <= 0:
        raise FpqError(f"normalize of zero/degenerate quaternion (norm2={n2})")
    r = rsqrt(n2)
    w, x, y, z = q
    return (_fg(_div(w * r, ONE)), _fg(_div(x * r, ONE)),
            _fg(_div(y * r, ONE)), _fg(_div(z * r, ONE)))


def vrotate(q, v, _div=_rdiv):
    """v' = v + w·t + u×t, t = 2(u×v) — the written rounding order IS the spec."""
    w, ux, uy, uz = _quat(q, "q")
    vx, vy, vz = _vec(v, "v")
    tx = _fg(2 * _div(uy * vz - uz * vy, ONE))
    ty = _fg(2 * _div(uz * vx - ux * vz, ONE))
    tz = _fg(2 * _div(ux * vy - uy * vx, ONE))
    return (_fg(vx + _div(w * tx, ONE) + _div(uy * tz - uz * ty, ONE)),
            _fg(vy + _div(w * ty, ONE) + _div(uz * tx - ux * tz, ONE)),
            _fg(vz + _div(w * tz, ONE) + _div(ux * ty - uy * tx, ONE)))


def qnlerp(p, q, t, _div=_rdiv):
    _fin(t, "t")
    if not (0 <= t <= ONE):
        raise FpqError(f"nlerp t={t} outside [0, ONE] — refused, never clamped")
    pq = _quat(p, "p")
    qq = _quat(q, "q")
    if qdot(pq, qq, _div) < 0:                     # shortest path, exact negation
        qq = tuple(-c for c in qq)
    li = ONE - t
    blend = tuple(_fg(_div(pc * li + qc * t, ONE)) for pc, qc in zip(pq, qq))
    return qnormalize(blend, _div)


# ---- the pinned battery (the corpus is a digest of these results) ---------------------
H = ONE // 2
Q3 = 3 * ONE // 4
T3 = ONE // 3

QUATS = ((ONE, 0, 0, 0),
         (ONE, H, 0, 0),
         (ONE, H, Q3, -T3),
         (3 * ONE, -2 * ONE, ONE, H))

VECS = ((ONE, 0, 0),
        (0, ONE, 0),
        (H, Q3, -ONE),
        (5 * ONE, -3 * ONE, 2 * ONE + 12345))

RSQRT_IN = (1, 2, ONE // 4, H, ONE, 2 * ONE, 3 * ONE, 10 * ONE,
            (1 << 48) + 7919, COMP_MAX)

NLERP_T = (0, ONE // 4, H, ONE)


def _ser(a):
    return int(a).to_bytes(8, "big", signed=True)      # FIELDFP serialization


def _battery(_div=_rdiv):
    """The fixed op battery. Returns ordered (tag, [raw…]) rows — the corpus."""
    rows = []
    for x in RSQRT_IN:
        rows.append(("rsqrt", [rsqrt(x)]))
    for q in QUATS:
        rows.append(("norm2", [qnorm2(q, _div)]))
    for p in QUATS:
        for q in QUATS:
            rows.append(("qmul", list(qmul(p, q, _div))))
    units = []
    for q in QUATS:
        u = qnormalize(q, _div)
        units.append(u)
        rows.append(("normalize", list(u)))
    for u in units:
        for v in VECS:
            rows.append(("rotate", list(vrotate(u, v, _div))))
    for i in range(len(QUATS)):
        p, q = QUATS[i], QUATS[(i + 1) % len(QUATS)]
        for t in NLERP_T:
            rows.append(("nlerp", list(qnlerp(p, q, t, _div))))
    return rows


def battery_digest(_div=_rdiv):
    h = hashlib.sha256()
    h.update(MAGIC)
    for tag, vals in _battery(_div):
        h.update(tag.encode("ascii"))
        for v in vals:
            h.update(_ser(v))
    return h.hexdigest()


def battery_digest_defect_truncdiv():
    """Floor division instead of the frozen round-to-nearest — MUST diverge
    (the same law the field stage proves: truncation ≠ _rdiv)."""
    def flo(p, d):
        return p // d
    return battery_digest(_div=flo)


def battery_digest_defect_wrap64():
    """THE PORT DEFECT actually observed in this repo's history (rigidity port
    note: 'a naive i64 multiply wraps and diverges'): quaternion products and
    norm sums computed in wrapping 64-bit arithmetic instead of 128-bit. MUST
    diverge from the golden — the battery's large quaternion exists to force it."""
    def qnorm2_w(q):
        w, x, y, z = q
        s = _w64(_w64(_w64(w * w) + _w64(x * x)) + _w64(_w64(y * y) + _w64(z * z)))
        return _rdiv(s, ONE)

    def qmul_w(p, q):
        pw, px, py, pz = p
        qw, qx, qy, qz = q

        def m(a, b):
            return _w64(a * b)
        return (_rdiv(_w64(_w64(m(pw, qw) - m(px, qx)) - _w64(m(py, qy) + m(pz, qz))), ONE),
                _rdiv(_w64(_w64(m(pw, qx) + m(px, qw)) + _w64(m(py, qz) - m(pz, qy))), ONE),
                _rdiv(_w64(_w64(m(pw, qy) - m(px, qz)) + _w64(m(py, qw) + m(pz, qx))), ONE),
                _rdiv(_w64(_w64(m(pw, qz) + m(px, qy)) - _w64(m(py, qx) - m(pz, qw))), ONE))

    h = hashlib.sha256()
    h.update(MAGIC)
    for x in RSQRT_IN:
        h.update(b"rsqrt")
        h.update(_ser(rsqrt(x)))
    for q in QUATS:
        h.update(b"norm2")
        h.update(_ser(_w64(qnorm2_w(q))))
    for p in QUATS:
        for q in QUATS:
            h.update(b"qmul")
            for v in qmul_w(p, q):
                h.update(_ser(_w64(v)))
    units = [qnormalize(q) for q in QUATS]
    for u in units:
        h.update(b"normalize")
        for c in u:
            h.update(_ser(c))
    for u in units:
        for v in VECS:
            h.update(b"rotate")
            for c in vrotate(u, v):
                h.update(_ser(c))
    for i in range(len(QUATS)):
        p, q = QUATS[i], QUATS[(i + 1) % len(QUATS)]
        for t in NLERP_T:
            h.update(b"nlerp")
            for c in qnlerp(p, q, t):
                h.update(_ser(c))
    return h.hexdigest()


if __name__ == "__main__":
    d = battery_digest()
    rows = _battery()
    print(f"URDRFPQ1 battery: {len(rows)} rows, digest {d}")
    print("defect truncdiv diverges:", battery_digest_defect_truncdiv() != d)
    print("defect wrap64 diverges:", battery_digest_defect_wrap64() != d)
    print("isqrt ≡ math.isqrt spot:", all(_isqrt_newton(n) == math.isqrt(n)
                                          for n in (0, 1, 2, 3, 4, 5, 24, 25, 26,
                                                    (1 << 96) // ONE, (1 << 96) // 3)))
